### PR TITLE
[FIX] Cropping with strict boundaries.

### PR DESCRIPTION
**Pull request by jaeilepp**
Closes https://github.com/mne-tools/mne-python/issues/2825
Cropping used relative tolerance when determining the boundaries for the new data.

### CODE REVIEW

**Comment by agramfort** [comment]
can you explain what the problem was?

**Comment by jaeilepp** [comment]
``_time_mask`` uses ``numpy.isclose`` to include nearby samples. It uses relative tolerances, so when the values are big, more samples are included. I'm not sure where this 'non strict' behavior is desired.

**Comment by agramfort** [comment]
the non-strict mode seems buggy...

**Comment by jaeilepp** [comment]
Ok, the other option would be to set the ``rtol=0``, in which case I think the crop would be inclusive. That's where the test is failing at the moment. I don't know if it will break some other stuff, but I think not. Which one do we want?

**Comment by jaeilepp** [comment]
Actually, I think this is related to the tmin tmax issue we've been having. I'll try the rtol option. I'll look deeper if it breaks something.

**Comment by agramfort** [comment]
Eric wake up ! :)

**Comment by Eric89GXL** [comment]
> It uses relative tolerances, so when the values are big, more samples are included. I'm not sure where this 'non strict' behavior is desired.

I'm not sure much we can do about this -- I think any solution has potential drawbacks. Floating point arithmetic is not perfect, so you need both an absolute tolerance so points near zero e.g. 0. and 1e-12 are considered the same, and large points that differ by numerical epsilon are also considered the same (even when dealing with potential multiplies, divides, etc. and saving/loading in `float32` format). It is true that with our current `master` solution using `isclose`, given the fixed sampling rates we use, when values get very big relative to the sample rate, more samples will be included by our use of `isclose`. The alternative -- not using isclose and relying on floating point arithmetic to land us on the correct side of the comparator -- by contrast will likely have the problem that at any sample rate and any number of samples, we might not be able to guarantee which samples will be included, and which won't. I think the latter problem is worse than the former.

What problem did you actually hit with this in practice? Did you have data sampled at 1000 Hz and find that once you got into the 10,000's of seconds, more data points than you expected were included? Maybe we should just expose the `strict` argument to the user, in which case both the sample-rounding and `isclose` behavior will not be used.

**Comment by Eric89GXL** [comment]
Ahh, it's from this:

https://github.com/mne-tools/mne-python/issues/2825#issuecomment-176104168

The inclusion of more points actually does make sense to me in this context, assuming we want certain behaviors in our cropping functions. I think we need to decide which tradeoffs we want to accept, and which we don't. This is the third time we've tried to sort it out now: once with the initial addition of `_time_mask` (#1855, #1811), then with the recent changes (#2802, #2787), and now with this PR (#2826, #2825).

Maybe with our new sample-based rounding we *can* get away with not using `isclose` at all, if we're really careful about how we compute our `times` (i.e., always using int64 divided by float64 `sfreq` in `info`). But I'm skeptical.

**Comment by agramfort** [comment]
I probably miss something but to me the logic is just to say that if times
is

1, 2, 3, 4

if tmin is 2.2 we include 2 as 2.2 is closer to 2 than to 3

what am I missing?

**Comment by jaeilepp** [comment]
Yes, the problem is that with floating points you cannot always be sure which one is closer. I think we can at least somehow isolate the problem to some certain functions like ``decimate.resample``. Maybe there should be a warning if the user tries to decimate with something else than power of two values.

**Comment by Eric89GXL** [comment]
> if tmin is 2.2 we include 2 as 2.2 is closer to 2 than to 3
> what am I missing?

Nothing in theory. But in practice we need to make sure it works for ugly cases like the MGH default sample rate, and that when we multiply and divide out to go to a sample number and back to a time, that we still include the correct sample when using strict (non-`isclose`) comparison. And this needs to still work -- i.e., include the same set of samples -- when there are multiple steps of I/O involved with casting times, `tmin`, `tstep` (for the `SourceEstimate` case), or `sfreq` (for most everything else) to `float32`, writing to disk, reading back, and then redoing all the comparisons. `isclose` has made it so that we haven't needed to worry about getting things exactly right and consistent in a lot of places, and getting rid of it means we'll need to be more careful.

**Comment by agramfort** [comment]
yeah but isclose is basically broken too as shown here ...

**Comment by Eric89GXL** [comment]
It's possible that we can get closer to behavior people would expect by getting rid of `isclose` now that we (try to) push the tmin/tmax to the nearest sample equivalent. But someone will need to put in the time to try to get it right in as many corner cases as possible. @jaeilepp have time to try getting rid of `isclose` completely and see if you can make it work?

**Comment by jaeilepp** [comment]
Ok, tests seem to pass now, but I'm not super confident I didn't break something in the process. Feel free to test it out. I'll take a closer look on monday.

File: mne/utils.py [hunk 1]
```diff
@@ -1872,17 +1872,23 @@ def _time_mask(times, tmin=None, tmax=None, strict=False, sfreq=None):
...
+            else:
+                tmin = times[0]
             if np.isfinite(tmax):
                 tmax = round(tmax * sfreq) / sfreq
-        mask |= isclose(times, tmin)
-        mask |= isclose(times, tmax)
+            else:
+                tmax = times[-1]
+        deltas = np.abs(times - tmax)  # Find nearest times
+        tmax = times.flat[np.where(deltas == deltas.min())[0]][-1]
```
Comment (Eric89GXL) [inline mne/utils.py]: I think instead of doing `.flat` you should just guarantee above that `times` is `1D` (and change any calling functions to provide a 1D properly)

Shouldn't the min be `[0]` while the max should be `[-1]`? It seems like we'd want to push the min leftward and the max rightward in a "tie".

File: mne/viz/tests/test_topomap.py [hunk 2]
```diff
@@ -68,7 +68,7 @@ def test_plot_topomap():
 
     ev_bad = evoked.pick_types(meg=False, eeg=True, copy=True)
     ev_bad.pick_channels(ev_bad.ch_names[:2])
-    ev_bad.plot_topomap(times=ev_bad.times[:2] - 1e-6)  # auto, should plot EEG
+    ev_bad.plot_topomap(times=ev_bad.times[1:3] - 1e-6)  # auto, plots EEG
```
Comment (Eric89GXL) [inline mne/viz/tests/test_topomap.py]: So your changes modify which points are included here? It seems like it shouldn't.

File: mne/viz/topomap.py [hunk 3]
```diff
@@ -1126,7 +1126,9 @@ def plot_evoked_topomap(evoked, times="auto", ch_type=None, layout=None,
         times = _process_times(evoked, times, n_peaks=len(axes))
     else:
         times = _process_times(evoked, times, n_peaks=None)
-
+    if max(times) > evoked.times[-1] or min(times) < evoked.times[0]:
+        raise ValueError('Times should be between {0:0.3f} and '
+                         '{1:0.3f}.'.format(evoked.times[0], evoked.times[-1]))
```
Comment (Eric89GXL) [inline mne/viz/topomap.py]: I suspect this will break something that it shouldn't. For example with MGH sample rate I think you can set a `tmin` in between two samples that's closer to sample above it than below it, then if you pass that `tmin` in as a time, it will throw an error here that it wouldn't before. Is this why you had to modify the test above?

File: mne/cov.py [hunk 4]
```diff
@@ -51,8 +51,7 @@ def _check_covs_algebra(cov1, cov2):
 
 def _get_tslice(epochs, tmin, tmax):
     """get the slice."""
-    tstart, tend = None, None
-    mask = _time_mask(epochs.times, tmin, tmax, sfreq=epochs.info['sfreq'])
+    mask = _time_mask(epochs.times, tmin, tmax, strict=True)
```
Comment (Eric89GXL) [inline mne/cov.py]: Now `cov` and `SourceEstimate` calculations are the two places where `strict=True` is used, and `strict=False` is used everywhere else? That doesn't seem consistent to me.

File: mne/viz/tests/test_topomap.py (hunk 2 above)
Comment (agramfort) [inline mne/viz/tests/test_topomap.py]: +1. @jaeilepp any idea why?

File: mne/cov.py (hunk 4 above)
Comment (agramfort) [inline mne/cov.py]: I think we should use strict=False everywhere and get rid of the strict parameter.

Comment (jasmainak) [inline mne/cov.py]: +1 on that

Comment (jaeilepp) [inline mne/cov.py]: Yeah, strict was only used on some operation on ``SourceEstimate``. Didn't dare change it. Will check monday..

File: mne/utils.py (hunk 1 above)
Comment (jaeilepp) [inline mne/utils.py]: Hmm. How about rounding upwards on a tie? I think it could work, although this is kind of difficult to get my head around the whole thing with floats.

File: mne/viz/tests/test_topomap.py (hunk 2 above)
Comment (jaeilepp) [inline mne/viz/tests/test_topomap.py]: The checker for topomap is now more strict. It doesn't allow any values outside the ``times`` window, so this test would fail. The checker had to change because ``_time_mask`` is more lenient now in allowing values to be anything, not just close to ``times``. In my opinion we could get rid of the checker for topomap altogether, but I think some people want it there.

File: mne/viz/topomap.py (hunk 3 above)
Comment (jaeilepp) [inline mne/viz/topomap.py]: Yeah, in my opinion we could get rid of this check.

File: mne/viz/tests/test_topomap.py (hunk 2 above)
Comment (agramfort) [inline mne/viz/tests/test_topomap.py]: you should loosen the check. Values should be between

tmin - 1. / (2. * sfreq) and tmax + 1 /( 2. * sfreq)

File: mne/viz/topomap.py (hunk 3 above)
Comment (agramfort) [inline mne/viz/topomap.py]: just make it less strict

**Comment by jaeilepp** [comment]
Ready for review. Strict is no longer used anywhere.

**Comment by agramfort** [comment]
travis is not happy

**Comment by Eric89GXL** [comment]
LGTM, but could you add a test along the lines of #2825? You could make a `RawArray` with a single channel @ 1000 Hz and 10k samples and you'll probably see the problem appear. Please make sure it shows the bad behavior on `master` but fixes it here.

File: mne/beamformer/tests/test_lcmv.py [hunk 5]
```diff
@@ -83,7 +83,7 @@ def _get_data(tmin=-0.1, tmax=0.15, all_forward=True, epochs=True,
                                    eeg=0.1, proj=True)
     if data_cov:
         with warnings.catch_warnings(record=True):
-            data_cov = mne.compute_covariance(epochs, tmin=0.04, tmax=0.15)
+            data_cov = mne.compute_covariance(epochs, tmin=0.04, tmax=0.145)
```
Comment (agramfort) [inline mne/beamformer/tests/test_lcmv.py]: can you explain the impact of this change? what was failing?

Comment (jaeilepp) [inline mne/beamformer/tests/test_lcmv.py]: When it was changed to non-strict time_mask, it was picking one sample more. This needed to change to reflect that.

**Comment by agramfort** [comment]
besides LGTM

File: mne/beamformer/tests/test_lcmv.py (hunk 5 above)
Comment (agramfort) [inline mne/beamformer/tests/test_lcmv.py]: how bad does it affect the tests?

Comment (jaeilepp) [inline mne/beamformer/tests/test_lcmv.py]: I think it ended up failing here https://github.com/mne-tools/mne-python/blob/master/mne/beamformer/tests/test_lcmv.py#L111

Comment (agramfort) [inline mne/beamformer/tests/test_lcmv.py]: how much does it affect the tmax?

Comment (jaeilepp) [inline mne/beamformer/tests/test_lcmv.py]: Very much. It was 0.65 or something.

Comment (Eric89GXL) [inline mne/beamformer/tests/test_lcmv.py]: IIRC the beamforming tests are a bit temperamental because they use very
few epochs, so I'm not surprised this needed to be tweaked

File: mne/io/fiff/tests/test_raw_fiff.py [hunk 6]
```diff
@@ -799,6 +799,14 @@ def test_crop():
         assert_equal(raw.last_samp, ar.last_samp)
         assert_array_equal(raw[:, :][0], ar[:, :][0])
 
+    # test shape consistency of cropped raw
+    data = np.zeros((1, 1002001))
+    info = create_info(['1'], 1000)
+    raw = RawArray(data, info)
+    for tmin in np.arange(0, 1001, 100):
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw_fiff.py]: FYI `range` should be a little bit better on Py3k because it's an iterator. I suspect it might even be faster in Py27. Any reason to use `arange` here instead of `range`? Not a blocker for merge but I'm curious.

File: mne/io/fiff/tests/test_raw_fiff.py [hunk 7]
```diff
@@ -799,6 +799,14 @@ def test_crop():
...
+        raw1 = raw.crop(tmin=tmin, tmax=tmin + 2, copy=True)
+        assert_equal(raw1._data.shape, (1, 2001))
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw_fiff.py]: don't use private attributes unless you need to, so `raw1[:][0].shape` would be better

File: mne/io/fiff/tests/test_raw_fiff.py [hunk 8]
```diff
@@ -799,6 +799,14 @@ def test_crop():
         assert_equal(raw.last_samp, ar.last_samp)
         assert_array_equal(raw[:, :][0], ar[:, :][0])
 
+    # test shape consistency of cropped raw
+    data = np.zeros((1, 1002001))
+    info = create_info(['1'], 1000)
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw_fiff.py]: I think you can just do `create_info(1, 1000)`

File: mne/io/fiff/tests/test_raw_fiff.py (hunk 7 above)
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw_fiff.py]: (sometimes it might be worth using private attributes for speed or memory reasons, but I don't think that's the case here...?)

**Comment by Eric89GXL** [comment]
@jaeilepp I'll go ahead and merge once the last round of minor edits are done, assuming you agree with the suggested changes

File: mne/io/fiff/tests/test_raw_fiff.py (hunk 6 above)
Comment (jaeilepp) [inline mne/io/fiff/tests/test_raw_fiff.py]: I actually thought ``step`` was only available in ``arange``. I'll change it.

Comment (agramfort) [inline mne/io/fiff/tests/test_raw_fiff.py]: I am happy as soon as Eric is happy :)

**Comment by Eric89GXL** [comment]
I'm happy as soon as Travis is happy :)

**Comment by Eric89GXL** [comment]
Thanks @jaeilepp
